observIQ exporter -  Conversion to observIQ format implementation

BinaryFissionGames opened the pull request:
**Description:**
Implementation of pdata.Logs to observIQ format.

**Link to tracking Issue:**
N/A

**Testing:**
Unit tests added for logdataToObservIQFormat
Tweaked unit tests for client to ensure request has the logs on the request in a parseable format

**Documentation:**
No (externally visible) changes to document

BinaryFissionGames committed:
observIQ exporter - implement transform to observIQ format

BinaryFissionGames committed:
dedot keys, add version in payload

BinaryFissionGames committed:
observIQ exporter - minor cleanup in converter.go

jpkrohling commented on exporter/observiqexporter/converter.go:
```diff
@@ -14,23 +14,36 @@
 
-import "go.opentelemetry.io/collector/model/pdata"
+import (
+	"crypto/md5" //nolint:gosec // Not used for crypto, just for generating ID based on contents of log
```
Is it required by your backend, or can it be anything? There are several hashing functions that could be appropriate here.

BinaryFissionGames replied:
MD5 isn't really a hard requirement, any hash function should work. I used MD5 here because it's what we use in our own agent, currently. If you have a suggestion for a hash function you think would fit better (maybe one of those BLAKE2 hash functions?), I'd be happy to hear it!

jpkrohling replied:
How about fnv? https://golang.org/pkg/hash/fnv/

BinaryFissionGames replied:
Good suggestion. Seems like it'll work just fine for our purpose, I've switched over to it in favor of md5 here.

jpkrohling commented on exporter/observiqexporter/converter.go:
```diff
@@ -43,6 +56,178 @@ type observIQLogEntry struct {
+			ill := ills.At(j)
+			logs := ill.Logs()
+			for k := 0; k < logs.Len(); k++ {
+				oiqLogEntry := resourceAndInstrmentationLogToEntry(resMap, logs.At(k), agentID, agentName, buildInfo)
```
```suggestion
				oiqLogEntry := resourceAndInstrumentationLogToEntry(resMap, logs.At(k), agentID, agentName, buildInfo)
```

jpkrohling commented on exporter/observiqexporter/converter.go:
```diff
@@ -43,6 +56,178 @@ type observIQLogEntry struct {
+	return &observIQLogBatch{Logs: sliceOut}, errorsOut
+}
+
+// Output timestamp format, an ISO8601 compliant timesetamp with millisecond precision
```
```suggestion
// Output timestamp format, an ISO8601 compliant timestamp with millisecond precision
```

jpkrohling commented on exporter/observiqexporter/converter.go:
```diff
@@ -43,6 +56,178 @@ type observIQLogEntry struct {
+// Output timestamp format, an ISO8601 compliant timesetamp with millisecond precision
+const timestampFieldOutputLayout = "2006-01-02T15:04:05.000Z07:00"
+
+func resourceAndInstrmentationLogToEntry(resMap interface{}, log pdata.LogRecord, agentID string, agentName string, buildInfo component.BuildInfo) *observIQLogEntry {
```
```suggestion
func resourceAndInstrumentationLogToEntry(resMap interface{}, log pdata.LogRecord, agentID string, agentName string, buildInfo component.BuildInfo) *observIQLogEntry {
```

BinaryFissionGames replied:
Good catch on these typos -- fixed these.

BinaryFissionGames commented on exporter/observiqexporter/converter_test.go:
```diff
@@ -15,14 +15,229 @@
+			false,
+		},
+		{
+			"Body is map",
```
The reason why I don't `[]byte` is because as far as I know, there's no way for that type to be represented on pdata.LogRecord -- the AttributeValue struct seems to only support null, string, int, double, bool, map, and array. Maybe this is more of a concern earlier in the pipeline? Otherwise, I'd need help figuring out how to represent this case.

djaglowski replied:
I see that it was added to the [proto](https://github.com/open-telemetry/opentelemetry-proto/pull/297/files), but hasn't yet been incorporated into the collector. We can worry about this later.

BinaryFissionGames commented on exporter/observiqexporter/client.go:
```diff
@@ -42,6 +43,7 @@ type client struct {
 	throttleTimer  *time.Timer
 	throttleLock   sync.RWMutex
 	timesThrottled int
+	buildInfo      component.BuildInfo
```
Yeah, good call. No use lugging around the whole struct if I'm only using the one piece.

BinaryFissionGames committed:
observiq exporter - fix some typos in converter.go

BinaryFissionGames committed:
observIQ exporter - client only needs version info, not all build info

BinaryFissionGames committed:
observIQ exporter - modify/add testing for converter.go

Modify test for various types on the resource attribs
Add test for when both body and attributes for the log

BinaryFissionGames commented on exporter/observiqexporter/converter_test.go:
```diff
@@ -15,14 +15,229 @@
+				logRecord.SetTimestamp(nanoTs)
+				return logRecord
+			},
+			pdata.NewResource,
```
Yeah, it can vary just as much, good point. Modified this test to also have all the types on the Resource.Attributes() as well.

BinaryFissionGames committed:
Let Body be anything, not just a map

Don't want to drop non-map/string values on body.

BinaryFissionGames commented on exporter/observiqexporter/converter_test.go:
```diff
@@ -15,14 +15,229 @@
+				Resource:  map[string]interface{}{},
+				Data:      nil,
+				Agent:     &observIQAgentInfo{Name: "agent", ID: "agentID", Version: "latest"},
+				Body: map[string]interface{}{
+					"mapKey": "value",
+				},
```
I added a test for this; Basically, current implementation is: 
Body (string) -> Message
Body (not a string) -> Body
Attributes -> Data (always)

I realize now that I had Body dropped if it wasn't a string/map -- I changed this and added some tests for other types of bodies (int, array).

BinaryFissionGames committed:
Use fnv hash over md5

BinaryFissionGames committed:
observIQ exporter - test hash length, fix bad hash usage

djaglowski reviewed (approved):
LGTM

@andykellr Do you mind having a look over the converter tests? Maybe kick the tires a bit?

bogdandrutu commented:
@djaglowski can I merge this? Please add ready-to-merge when ready.

@BinaryFissionGames please rebase

tigrannajaryan commented on exporter/observiqexporter/converter.go:
```diff
@@ -39,10 +51,193 @@ type observIQLogEntry struct {
+// Output timestamp format, an ISO8601 compliant timestamp with millisecond precision
+const timestampFieldOutputLayout = "2006-01-02T15:04:05.000Z07:00"
+
+func resourceAndInstrumentationLogToEntry(resMap interface{}, log pdata.LogRecord, agentID string, agentName string, buildVersion string) *observIQLogEntry {
```
Nit: I think `resMap` can be declared as more narrow and strong type `map[string]interface{}`.

tigrannajaryan commented on exporter/observiqexporter/converter.go:
```diff
@@ -39,10 +51,193 @@ type observIQLogEntry struct {
+	return ""
+}
+
+// If Body is not a string, it is suitable to be used on the observIQ log entry as "body"
```
Is this supposed to read?
```suggestion
// If Body is not a string, it is not suitable to be used on the observIQ log entry as "body"
```

tigrannajaryan reviewed (approved):
I'll merge. Feel free to address the comments in a separate PR if you want.
